into/backends/aws.py

```diff
@@ -0,0 +1,248 @@
+            bucket = self.s3.get_bucket(self.bucket)
+        except boto.exception.S3ResponseError:
+            bucket = self.s3.create_bucket(self.bucket)
+
+        self.object = bucket.get_key(self.key)
+        if self.object is None:
+            self.object = bucket.new_key(self.key)
+
+        self.subtype.__init__(self, uri, *args, **kwargs)
+
+
+@memoize
+def S3(cls):
+    """Parametrized S3 bucket Class
+
+    Examples
+    --------
+    >>> import sqlalchemy
+    >>> S3(CSV)
+    S3(CSV)
+    >>> S3(sqlalchemy.Table)
+    """
+    return type('S3(%s)' % cls.__name__, (_S3, cls), {'subtype': cls})
+
+
+@sample.register((S3(CSV), S3(JSONLines)))
+@contextmanager
+def sample_s3_line_delimited(data, length=8192):
+    """Get bytes from `start` to `stop` of `key` from the S3 bucket `bucket`.
+
+    Parameters
+    ----------
+    data : S3(CSV)
+        A CSV file living in  an S3 bucket
+    length : int, optional, default ``8192``
+        Number of bytes of the file to read
+
+    """
+    headers = {'Range': 'bytes=0-%d' % length}
+    raw = data.object.get_contents_as_string(headers=headers)
+
+    # this is generally cheap as we usually have a tiny amount of data
+    try:
+        index = raw.rindex(b'\r\n')
+    except ValueError:
+        index = raw.rindex(b'\n')
+
+    raw = raw[:index]
+
+    with tmpfile(ext(data.path)) as fn:
+        # we use wb because without an encoding boto returns bytes
+        with open(fn, 'wb') as f:
+            f.write(raw)
+        yield fn
+
+
+@discover.register((S3(CSV), S3(JSONLines)))
+def discover_s3_line_delimited(c, length=8192, **kwargs):
+    """Discover CSV and JSONLines files from S3.
+
+    Examples
+    --------
+    >>> csv_dshape = discover(resource('s3://nyqpug/tips.csv'))
+    >>> print(csv_dshape)
+    var * {
+      total_bill: ?float64,
+      tip: ?float64,
+      sex: ?string,
+      smoker: ?string,
+      day: ?string,
+      time: ?string,
+      size: int64
+      }
+    >>> json_dshape = discover(resource('s3://nyqpug/tips.json'))
+    >>> names = list(map(str, sorted(json_dshape.measure.names)))
+    >>> names
+    ['day', 'sex', 'size', 'smoker', 'time', 'tip', 'total_bill']
+    >>> types = [json_dshape.measure[name] for name in names]
+    >>> from pprint import pprint
+    >>> pprint(types)
+    [ctype("string"),
+     ctype("string"),
+     ctype("int64"),
+     ctype("string"),
+     ctype("string"),
+     ctype("float64"),
+     ctype("float64")]
+    """
+    with sample(c, length=length) as fn:
+        return discover(c.subtype(fn, **kwargs), **kwargs)
+
+
+@resource.register('s3://.*\.csv(\.gz)?', priority=18)
+def resource_s3_csv(uri, **kwargs):
+    return S3(CSV)(uri, **kwargs)
+
+
+@resource.register('s3://.*\.txt(\.gz)?', priority=18)
+def resource_s3_text(uri, **kwargs):
+    return S3(TextFile)(uri)
+
+
+@resource.register('s3://.*\.json(\.gz)?', priority=18)
+def resource_s3_json_lines(uri, **kwargs):
+    return S3(JSONLines)(uri, **kwargs)
```

Comment:
Thanks for extending the functions below to `TextFile`.  The ones above could use some love too.

If you can think of a way to cleanly avoid code duplication without too much extra complexity I'd be thrilled.  This has me stumped.